### PR TITLE
[TIR] Introduce ReduceBranchingThroughOvercompute

### PR DESCRIPTION
**Pull request by Lunderberg**
This PR introduces a new transform, which attempts to reduce branching by adding overcompute, where the overcompute has provably no effect on any observable value outside the PrimFunc.

### CODE REVIEW

**Comment by Lunderberg** [comment]
@csullivan This PR is ready for review and passing all unit tests, if you have a chance to review.

File: tests/python/unittest/test_tir_transform_simplify.py [hunk 1]
```diff
@@ -1689,5 +1689,26 @@ def expected(A: T.Buffer[1, "int32"]):
...
+#             if i != 0:
+#                 if i * i == 0:
+#                     A[i] = 5
+
+#     def expected(A: T.Buffer[16, "int32"]):
+#         for i in T.serial(16):
+#             if i != 0:
+#                 T.evaluate(0)
+
+
```
Comment (csullivan) [inline tests/python/unittest/test_tir_transform_simplify.py:1712]: Will need to remove this.

Comment (Lunderberg) [reply tests/python/unittest/test_tir_transform_simplify.py]: Thank you, and removed!

**Comment by csullivan** [comment]
@tvm-bot rerun

**Comment by LeiWang1999** [comment]
hi @Lunderberg , did we implement `Stmt Simplify(Stmt stmt, arith::Analyzer* analyzer);`?

**Comment by Lunderberg** [comment]
@LeiWang1999 We did, yes.  It's located [here](https://github.com/apache/tvm/blob/main/src/tir/transforms/simplify.h#L38).

That said, in most cases it's better to apply [`tir::transform::Simplify`](https://github.com/apache/tvm/blob/main/src/tir/transforms/simplify.h#L38) instead.  That way, the full context is available for a single simplification call.  Simplifying a statement in isolation may miss context-dependent simplifications.  (e.g. Within an `if i==0` block, `if i + 1 < 16` could be simplified to to `True`.)

**Pull request merged**